HLTrigger/special/plugins/HLTPPSPerPotTrackFilter.cc

```diff
@@ -0,0 +1,192 @@
+public:
+  explicit HLTPPSPerPotTrackFilter(const edm::ParameterSet&);
+  ~HLTPPSPerPotTrackFilter() override = default;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions&);
+  bool filter(edm::StreamID, edm::Event&, const edm::EventSetup&) const override;
+
+private:
+  edm::EDGetTokenT<edm::DetSetVector<CTPPSPixelLocalTrack>> pixelLocalTrackToken_;
+  edm::EDGetTokenT<edm::DetSetVector<TotemRPLocalTrack>> stripLocalTrackToken_;
+  edm::EDGetTokenT<edm::DetSetVector<CTPPSDiamondLocalTrack>> diamondLocalTrackToken_;
+
+  struct PerPotFilter {
+    int minTracks;
+    int maxTracks;
+  };
+  std::unordered_map<uint32_t, PerPotFilter> pixel_filter_;
+  std::unordered_map<uint32_t, PerPotFilter> strip_filter_;
+  std::unordered_map<uint32_t, PerPotFilter> diam_filter_;
+
+  // Helper tool to count valid tracks
+  static constexpr auto valid_trks_ = [](const auto& trk) { return trk.isValid(); };
+};
+
+void HLTPPSPerPotTrackFilter::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("pixelLocalTrackInputTag", edm::InputTag("ctppsPixelLocalTracks"))
+      ->setComment("input tag of the pixel local track collection");
+  desc.add<edm::InputTag>("stripLocalTrackInputTag", edm::InputTag("totemRPLocalTrackFitter"))
+      ->setComment("input tag of the strip local track collection");
+  desc.add<edm::InputTag>("diamondLocalTrackInputTag", edm::InputTag("ctppsDiamondLocalTracks"))
+      ->setComment("input tag of the diamond local track collection");
+
+  edm::ParameterSetDescription filterValid;
+  filterValid.add<unsigned int>("detid", 0)->setComment("station/pot raw DetId");
+  filterValid.add<int>("minTracks", -1)->setComment("minimum number of tracks in pot");
+  filterValid.add<int>("maxTracks", -1)->setComment("maximum number of tracks in pot");
+
+  std::vector<edm::ParameterSet> vPixelDefault;
+  auto& near_pix45 = vPixelDefault.emplace_back();
+  near_pix45.addParameter<unsigned int>("detid", CTPPSPixelDetId(0, 2, 2));  // arm-station-rp
+  near_pix45.addParameter<int>("minTracks", 2);
+  near_pix45.addParameter<int>("maxTracks", -1);
+  auto& far_pix45 = vPixelDefault.emplace_back();
+  far_pix45.addParameter<unsigned int>("detid", CTPPSPixelDetId(0, 2, 3));  // arm-station-rp
+  far_pix45.addParameter<int>("minTracks", 2);
+  far_pix45.addParameter<int>("maxTracks", -1);
+  auto& near_pix56 = vPixelDefault.emplace_back();
+  near_pix56.addParameter<unsigned int>("detid", CTPPSPixelDetId(1, 2, 2));  // arm-station-rp
+  near_pix56.addParameter<int>("minTracks", 2);
+  near_pix56.addParameter<int>("maxTracks", -1);
+  auto& far_pix56 = vPixelDefault.emplace_back();
+  far_pix56.addParameter<unsigned int>("detid", CTPPSPixelDetId(1, 2, 3));  // arm-station-rp
+  far_pix56.addParameter<int>("minTracks", 2);
+  far_pix56.addParameter<int>("maxTracks", -1);
+  desc.addVPSet("pixelFilter", filterValid, vPixelDefault);
```

Comment:
@makortel do you know if there is a more efficient (or less verbose) way to construct the `ParameterSet`, without having to add the parameters one by one ?

Reply:
I'm afraid there is not (as far as I can tell). The `edm::ParameterSet` does have a constructor from string, but that is in "encoded format" that I suppose is not (well) documented, and is difficult to decipher from the code alone.

Reply:
And since `edm::ParameterSet::addParameter` is void-returning (and not returning its own reference) I am not sure how to chain it better. Possibly I can try messing around with [`edm::ParameterSet::allToString`](https://github.com/cms-sw/cmssw/blob/master/FWCore/ParameterSet/interface/ParameterSet.h#L102) to find a possible string-tokenisable constructor for them, but I am afraid this might even worsen this function's readability...

Reply:
OK, no... I was wondering if there was a way to build a `ParameterSet` from a list of parameters (like one can build a `vector` from an initialiser list), but if that is not the case, anything more complicated may make the code harder to read instead of simpler :-)

Reply:
:ok_hand: